com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/nodes/altrep/AltrepDuplicateNode.java

```diff
@@ -0,0 +1,75 @@
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.r.runtime.data.NativeDataAccess.NativeMirror;
+import com.oracle.truffle.r.runtime.data.RIntArrayVectorData;
+import com.oracle.truffle.r.runtime.data.RIntVector;
+import com.oracle.truffle.r.runtime.data.altrep.AltIntegerClassDescriptor;
+import com.oracle.truffle.r.runtime.data.altrep.AltrepUtilities;
+import com.oracle.truffle.r.runtime.ffi.AltrepRFFI;
+import com.oracle.truffle.r.runtime.ffi.util.NativeMemory;
+import com.oracle.truffle.r.runtime.ffi.util.NativeMemory.ElementType;
+import com.oracle.truffle.r.runtime.nodes.RBaseNode;
+
+/**
+ * Note that the behavior of GNU-R's Rf_duplicate in duplicate.c is a bit weird: When the altrep vector
+ * does not have <code>Duplicate</code> method registered, then calling <code>Rf_duplicate</code> on such
+ * altrep vector creates an atomic vector, not an altrep vector.
+ */
+@GenerateUncached
+public abstract class AltrepDuplicateNode extends RBaseNode {
+    public abstract Object execute(Object altVec, boolean deep);
+
+    @Specialization(guards = {"isAltrep(altIntVec)", "hasDuplicateMethod(altIntVec)"}, limit = "1")
+    public Object doAltIntWithDuplicateMethod(RIntVector altIntVec, boolean deep,
+                                              @CachedLibrary("getDuplicateMethod(altIntVec)") InteropLibrary duplicateMethodInterop,
+                                              @Cached("createBinaryProfile()") ConditionProfile hasMirrorProfile,
+                                              @Cached("createBinaryProfile()") ConditionProfile duplicateReturnsNullProfile,
+                                              @Cached AltrepRFFI.AltIntDataptrNode dataptrNode,
+                                              @Cached AltrepLengthNode lengthNode) {
+        AltIntegerClassDescriptor descriptor = getAltIntDescriptor(altIntVec);
+        Object duplicatedObject = descriptor.invokeDuplicateMethodCached(altIntVec, deep, duplicateMethodInterop, hasMirrorProfile);
+        if (duplicateReturnsNullProfile.profile(duplicatedObject == null)) {
+            return doStandardDuplicate(altIntVec, deep, dataptrNode, lengthNode);
+        } else {
+            assert duplicatedObject instanceof NativeMirror; // TODO: debug this line
+            assert ((NativeMirror) duplicatedObject).getDelegate() instanceof RIntVector;
+            RIntVector duplicatedVector = (RIntVector) ((NativeMirror) duplicatedObject).getDelegate();
+            // We have to return data, not the whole vector. However, the Duplicate method returns whole vector.
+            return duplicatedVector.getData();
+        }
+    }
+
+    @Specialization(guards = {"isAltrep(altIntVec)"}, replaces = {"doAltIntWithDuplicateMethod"}, limit = "3")
```

Comment:
Re `replaces`: it looks like you're OK with ignoring the duplicate method if the cache overflows, is this intentional? The cache probably never overflows given how it is used, but it is not intuitive